merlion/models/forecast/deep_base.py

```diff
@@ -0,0 +1,261 @@
+        **kwargs,
+    ):
+        """
+        :param n_past: # of past steps used for forecasting future.
+        :param start_token_len: Length of start token for deep transformer encoder-decoder based models such as
+            ``Autoformer`` or ``informer``. The start token is similar to the special tokens for NLP models (e.g., bos, sep, eos tokens).
+            For non-transformer based models, we set the ``start_token_len = 0``.
+        """
+        super().__init__(
+            **kwargs,
+        )
+        self.n_past = n_past
+        self.start_token_len = start_token_len
+
+
+class DeepForecaster(DeepModelBase, ForecasterBase):
+    """
+    Base class for a deep forecaster model
+    """
+
+    config_class = DeepForecasterConfig
+
+    def __init__(self, config: DeepForecasterConfig):
+        super().__init__(config)
+
+    def _get_np_loss_and_prediction(self, eval_dataset: RollingWindowDataset):
+        """
+        Get numpy prediction and loss with evaluation mode for a given dataset or data
+
+        :param evaluate_data: Evaluation data or dataset
+
+        :return: The numpy prediction of the model and the average loss for the given dataset.
+        """
+        config = self.config
+        self.deep_model.eval()
+
+        all_preds = []
+        all_trues = []
+        total_loss = []
+
+        for i, batch in enumerate(eval_dataset):
+            with torch.no_grad():
+                loss, outputs, y_true = self._get_batch_model_loss_and_outputs(self._convert_batch_to_tensors(batch))
+                pred = outputs.detach().cpu().numpy()
+                true = y_true.detach().cpu().numpy()
+
+                all_preds.append(pred)
+                all_trues.append(true)
+                total_loss.append(loss.item())
+
+        preds = np.concatenate(all_preds, axis=0)
+        trues = np.concatenate(all_trues, axis=0)
+
+        return preds, np.average(total_loss)
+
+    @property
+    def support_multivariate_output(self) -> bool:
+        """
+        Deep models support multivariate output by default.
+        """
+        return True
+
+    def _convert_batch_to_tensors(self, batch):
+        device = self.deep_model.device
+
+        past, past_timestamp, future, future_timestamp = batch
+
+        past = torch.tensor(past, dtype=torch.float, device=device)
+        future = future if future is None else torch.tensor(future, dtype=torch.float, device=device)
+
+        past_timestamp = torch.tensor(past_timestamp, dtype=torch.float, device=device)
+        future_timestamp = torch.tensor(future_timestamp, dtype=torch.float, device=device)
+
+        return (past, past_timestamp, future, future_timestamp)
+
+    def _train(self, train_data: pd.DataFrame, train_config=None) -> pd.DataFrame:
+        config = self.config
+
+        # creating model before the training
+        self._create_model()
+        if config.use_gpu:
+            self.to_gpu()
+        else:
+            self.to_cpu()
```

Comment:
Can you initialize the model on the appropriate device in `self._create_model()` only?